Add createwallet "disableprivatekeys" option: a sane mode for watchonly-wallets

jonasschnelli opened the pull request:
This mode ('createwallet {"disableprivatekeys": true}') is intended for a sane pure watch-only mode, ideal for a use-case where one likes to use Bitcoin-Core in conjunction with a hardware-wallet or another solutions for cold-storage.

Since we have support for custom change addresses in `fundrawtransaction`, pure watch-only wallets including coin-selection are possible and do make sense for some use cases.

This new mode disables all forms of private key generation and ensure that no mix between hot and cold keys are possible.

jonasschnelli committed:
[Qt] Disable creating receive addresses when private keys are disabled

TheBlueMatt commented:
It seems somewhat strange that this is set per-run and not per-wallet. Indeed, if it were set per-wallet, an easy way to get the intended behavior here is to simply encrypt your wallet with a garbage passphrase such that it can no longer be opened.

laanwj commented:
Concept ACK. I agree that this would be better as a mode on the wallet instead of yet another startup option.

NicolasDorier commented:
I would like to rebase on top of that for https://github.com/bitcoin/bitcoin/pull/9728

jonasschnelli committed:
Add facility to store wallet flags (64 bits)

jonasschnelli commented:
Added two commit.
* First commit adds a facility to store and check wallet flags (64bit). Made DISABLE_PRIVATE_KEYS the first flag.
* Second commit makes the disablehot function per wallet (no longer a global state)

Disabling private keys on a wallet that already contains private keys is not possible (also the other way around).

jonasschnelli commented on qa/rpc-tests/disablehot.py:
```diff
@@ -0,0 +1,85 @@
+        print("restarting node with disablehot=0 (try to re-enable private keys which must not be possible)")
+        stop_nodes(self.nodes)
+        try:
+            self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, [[], ['-disablehot=0']])
```
@MarcoFalke: any idea how I can make this pass? Right now, staring bitcoind in this case with this setup will cause bitcoind to write to stderr (and halt) which then result in pass=false in `rpc-tests.py`. But it's actually the correct behaviour.

MarcoFalke replied:
We might create a `assert_start_raises_init_error()` method. I will take a look at this tomorrow.

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -1408,13 +1407,25 @@ bool CWallet::IsHDEnabled()
 
+void CWallet::SetWalletFlag(uint64_t flag)
+{
+    walletFlags |= flag;
```
Kind of confusing....
I thought this would be equivalent to `CWallet::SetWalletFlags(uint64_t flags, false)`.

 However  while `CWallet::SetWalletFlags(uint64_t flags, bool memonly)` set the flags to the indicated value, `CWallet::SetWalletFlag(uint64_t flag)` **appends** to it.

Maybe you should rename to `AppendWalletFlags`.

saleemrashid replied:
`CWallet::AddWalletFlags(uint64_t flags, book memonly)` seems most consistent.

jonasschnelli replied:
IMO set flags seems most appropriate here. Flags are not getting *appended*. Either you set or unset IMO. For consistency, we should have a `UnsetWalletFlag()`.
But I agree its kind of confusing that we have a `SetWalletFlags` (plural) method that is capable or re-setting all 64 flags.

saleemrashid replied:
I think this method should be plural, since there is no distinction between a single flag and multiple flags in its operation and it might be useful to be able to do the latter.

Then, of course, you have confusion. So, a verb change in the method name might be useful.

NicolasDorier replied:
@jonasschnelli the unlogic thing is that [SetWalletFlag](https://github.com/bitcoin/bitcoin/pull/9662#discussion-diff-101948618R1410) switch a bit, when [SetWalletFlags](https://github.com/jonasschnelli/bitcoin/blob/e581d843f06d5fe71a4662b122f94577e5dccd55/src/wallet/wallet.cpp#L1417) set the walletFlags to a specific value.

jonasschnelli replied:
Maybe renaming `SetWalletFlags` (plural) to `OverwriteWalletFlags`?

NicolasDorier commented on qa/rpc-tests/disablehot.py:
```diff
@@ -0,0 +1,85 @@
+        self.num_nodes = 2
+
+    def setup_network(self, split=False):
+        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, [[], ['-disablehot']])
```
does the `[]` arguments is on purpose? because I think it makes `-disablehot` ignored

jonasschnelli replied:
We start two nodes here, the first node without startup parameters (thats why there is the `[]`). Only second node has `-disablehot`.

jonasschnelli commented:
Needed rebase (#8775).
Travis will fail because of the missing `assert_start_raises_init_error` (see #9832).

laanwj commented:
> ideal for a use-case where one likes to use Bitcoin-Core in conjunction with a hardware-wallet or another solutions for cold-stogare

Nice, so this guarantees that bitcoind process will contain no private key data at all?
Concept ACK.

Travis should pass now that #9832 is merged.

jonasschnelli commented:
The `disablehot.py` now makes use of the new (#9832) `assert_start_raises_init_error` call.

> Nice, so this guarantees that bitcoind process will contain no private key data at all?

Yes. Event the default key is disabled (while still detecting the first start correctly).

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -3670,6 +3715,27 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
+    else if (IsArgSet("-disablehot")) {
+        bool disableHot = GetBoolArg("-disablehot", DEFAULT_DISABLE_HOT_WALLET);
+        LOCK(walletInstance->cs_wallet);
+        if (walletInstance->vchDefaultKey.IsValid() && disableHot) {
```
I think you should use `fFirstRun` here instead of `walletInstance->vchDefaultKey.IsValid()`

luke-jr replied:
This should at least be abstracted somehow. vchDefaultKey's existence is not exactly identical to "does this wallet have keys?"

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -3670,6 +3715,27 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
+    else if (IsArgSet("-disablehot")) {
+        bool disableHot = GetBoolArg("-disablehot", DEFAULT_DISABLE_HOT_WALLET);
+        LOCK(walletInstance->cs_wallet);
+        if (walletInstance->vchDefaultKey.IsValid() && disableHot) {
```
At this point `fFirstRun` is always `false`. And I think this also works in conjunction with `salvagewallet`.

NicolasDorier replied:
Indeed. Might be out of this PR matter, in the case of hdwatchonly, `walletInstance->vchDefaultKey.IsValid()` will be true. So this would mean `disabledHot` is incompatible with hdwatchonly mode ?

NicolasDorier replied:
OK forget what I said, there is no reason to have disabledHot with hdwatchonly at same time.

jonasschnelli commented:
Rebased.

laanwj commented:
Needs rebase again (sorry)
utACK otherwise.

jonasschnelli commented:
Rebased.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3737,7 +3754,20 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
+        }
+        bool hotKeys = walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS);
+        if (hotKeys && !disableHot) {
+            InitError(strprintf(_("Error loading %s: You can't enable hot keys once you have initialized a wallet with disabled hot keys"), walletFile));
```
> Add `-disablehot` mode: a sane mode for watchonly-wallets

Maybe the error message could suggest passing `-disablehot` to avoid the problem. I think it would make the context of the error easier to understand for someone not expecting it.

ryanofsky reviewed (commented):
utACK 9b39bf94f168ab82707df91f21cffea0b286f919

The commit history of this PR is messy though. I would suggest moving the "Add facility to store wallet flags (64 bits)" commit to be first, then then squashing all the other changes down into single commit following that.

jonasschnelli committed:
Add option to disable private keys during internal wallet creation

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -2497,6 +2500,11 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                         //  post-backup change.
 
                         // Reserve a new key pair from key pool
+                        if (fDisableHotKeys) {
```
Falling earlier would disallow creating watch-only transactions (over fundrawtransaction).

jonasschnelli commented:
- Completely rewrote the commit history (as reported by @ryanofsky it was messy).
- Fixed @ryanofsky points

No strong opinion if this should be called `-disableprivatekeys` (which somehow nails it but may confuse people who just want cold key storage). Other opinions?

ryanofsky commented:
utACK fcea55a6ec6be57ee96458bde6fb24656269d3ee

Changes from previous review: pwalletMain removal, SetWalletFlag rename, error message tweaks, assert_raises_jsonrpc usage, and history cleanup.

NicolasDorier commented:
utACK fcea55a6ec6be57ee96458bde6fb24656269d3ee

ryanofsky commented:
This has 3 utACKs so I'm building it now to provide a little testing. Maybe this PR should also update `doc/release-notes.md` since it's adding a new option.

On "private keys" vs "hot keys" naming, "private keys" still sounds more comprehensible to me, but I wouldn't weigh my opinion very heavily if "hot key" is a term users will know.

ryanofsky commented:
Tested ACK fcea55a6ec6be57ee96458bde6fb24656269d3ee. Confirmed disablehot option prevents generating keys, persists across reload even when disablehot option not specified in later runs, and that it prevents loading pre-existing or non-disablehot wallets. I don't think I encountered any bugs, and it seems to me this PR works well and is safe to merge.

I was surprised by two things, and maybe these behaviors could be changed in the future if they aren't intentional:

- I was surprised that -disablehot could only be used on a new wallet and would refuse to load a preexisting wallet that did not have any private keys.
- I was surprised that -disablehot did seem to allow importing private keys via importprivkey and importwallet rpcs.

jonasschnelli commented:
Added two commits.
The first addresses the "importprivkey" / "Importwallet" issues reported by @ryanofsky (disable both commands in -disablehot mode).
The second fixes a rebase `pwallet`/`pwalletmain` issue.

ryanofsky reviewed (commented):
utACK 9e94cc41b3180949011d2299da14c13471b6aa2b. Did not retest, but the two new commits are straightforward and look good. This needs rebase, and the last pwalletMain fixing commit should probably be squashed into the main "Add per wallet -disablehot" comment.

instagibbs commented:
needs rebase, will review

instagibbs commented:
General comment while I wait for rebase: This is fundamentally in conflict with the hdwatchonly mode proposed in https://github.com/bitcoin/bitcoin/pull/9728 which was surprising to me. Perhaps the descriptions of the mode should stress that no (pub)key generation of any sort is allowed, or explicitly state that only imported watchonly keys are supported?

Also, I think to make this mode less brittle if an assert in `AddKeyPubKey` is added, since this should never be reachable.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -2833,6 +2839,9 @@ DBErrors CWallet::LoadWallet(bool& fFirstRunRet)
     if (nLoadWalletRet != DB_LOAD_OK)
         return nLoadWalletRet;
     fFirstRunRet = !vchDefaultKey.IsValid();
+    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS)) {
```
Agree, and combine the two conditionals to make it clearer that that's what is happening.

```
fFirstRunRet = !vchDefaultKey.IsValid() && !IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS);
```

luke-jr replied:
Why are we touching this logic at all? Seems we should just set a default key for watchonly-only wallets, until default keys are removed normally...

jnewbery commented on test/functional/disablehot.py:
```diff
@@ -0,0 +1,67 @@
+# Copyright (c) 2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
Can you add a short docstring explaining what the test is testing and how it is testing?

jnewbery commented on test/functional/disablehot.py:
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
```
Consider running a linter like flake8 over python modules to check for common nits.

jnewbery reviewed (commented):
Generally looks good. A few comments inline. I haven't tested yet other than running the new functional test. I'll do some manual testing tomorrow.

I like the flag infrastructure you've put in place. I wonder if it can be used in #7729 as a way to upgrade from accounts to labels.

It'd be good to add the flags to the `getwalletinfo` output, but that can be done in a future PR.

I have a very slight preference for naming this `-disableprivatekeys`, but `-disablehot` is also fine.

luke-jr commented on src/wallet/walletdb.cpp:
```diff
@@ -537,6 +537,12 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+        {
+            uint64_t flags;
+            ssValue >> flags;
+            pwallet->SetWalletFlags(flags, true);
```
I would think we should fail somehow, if an unrecognised flag is set?

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -3598,6 +3617,7 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
 {
     std::string strUsage = HelpMessageGroup(_("Wallet options:"));
     strUsage += HelpMessageOpt("-disablewallet", _("Do not load the wallet and disable wallet RPC calls"));
+    strUsage += HelpMessageOpt("-disablehot", _("Disable the possibility of hot keys (only watchonlys are possible in this mode") + " " + strprintf(_("(default: %u)"), DEFAULT_DISABLE_HOT_WALLET));
```
Prefer:

```C++
strprintf(_("Disable the possibility of hot keys (only watchonlys are possible in this mode; default: %u)"), DEFAULT_DISABLE_HOT_WALLET)
```

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -3728,6 +3751,27 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
+        }
+        bool hotKeys = walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS);
+        if (hotKeys && !disableHot) {
+            InitError(strprintf(_("Error loading %s: You can't enable hot keys once you have initialized a wallet with disabled hot keys (use -disablehot instead)"), walletFile));
```
Drop the "`(use -disablehot instead)`", since that will also fail in a multiwallet scenario.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -3717,6 +3737,9 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
                 throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
         }
+        if (GetBoolArg("-disablehot", DEFAULT_DISABLE_HOT_WALLET)) {
+            walletInstance->AddWalletFlag(WALLET_FLAG_DISABLE_HOT_KEYS);
```
This should also probably fake an encrypted wallet, so it can be read by older versions sanely.

luke-jr commented on src/wallet/rpcdump.cpp:
```diff
@@ -102,6 +102,10 @@ UniValue importprivkey(const JSONRPCRequest& request)
 
 
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: Hot keys are disabled (-disablehot)");
```
It doesn't make sense to mention the option here, since it's tied to the wallet and can't just be toggled...

luke-jr commented on src/wallet/rpcdump.cpp:
```diff
@@ -470,6 +474,10 @@ UniValue importwallet(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_WALLET_ERROR, "Importing wallets is disabled in pruned mode");
 
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: Hot keys are disabled (-disablehot)");
```
It doesn't make sense to mention the option here, since it's tied to the wallet and can't just be toggled...

promag commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -93,6 +93,10 @@ void ReceiveCoinsDialog::setModel(WalletModel *_model)
         // Last 2 columns are set by the columnResizingFixer, when the table geometry is ready.
         columnResizingFixer = new GUIUtil::TableViewLastColumnResizingFixer(tableView, AMOUNT_MINIMUM_COLUMN_WIDTH, DATE_COLUMN_WIDTH, this);
+
+        if (_model->hotKeysDisabled()) {
```
```cpp
ui->receiveButton->setEnabled(!_model->hotKeysDisabled());
```

promag commented on src/qt/walletmodel.cpp:
```diff
@@ -703,6 +703,11 @@ bool WalletModel::hdEnabled() const
 
+bool WalletModel::hotKeysDisabled() const
+{
+    return (wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_HOT_KEYS));
```
Drop outer `()`?

promag commented on src/wallet/walletdb.cpp:
```diff
@@ -871,6 +877,12 @@ bool CWalletDB::WriteHDChain(const CHDChain& chain)
     return batch.Write(std::string("hdchain"), chain);
 }
 
+bool CWalletDB::WriteWalletFlags(const uint64_t flags)
```
Remove const.

promag commented on src/wallet/walletdb.h:
```diff
@@ -212,6 +212,7 @@ class CWalletDB
     //! write the hdchain model (external chain child index counter)
     bool WriteHDChain(const CHDChain& chain);
 
+    bool WriteWalletFlags(const uint64_t flags);
```
Remove const.

kallewoof replied:
Why?

promag commented on test/functional/disablehot.py:
```diff
@@ -0,0 +1,67 @@
+
+class DisableHotKeysTest(BitcoinTestFramework):
+
+    def __init__(self):
```
`set_test_params` instead.

promag commented:
Concept ACK. Partial review.

This option affects all loaded wallets? How will this interact with #10740?

promag reviewed (commented):
Concept ACK. Partial review. Needs rebase.

This option affects all loaded wallets? How will this interact with #10740?

Sjors commented:
Needs rebase.

Tend to agree that this should be a per wallet thing and that this should be set upon wallet creation, rather than application launch.

If I understand correctly this is purely a safety feature. But I wonder if it doesn't make too many assumptions.

The term watch-only wallet is ambiguous. It could mean that the user doesn't have access to private keys at all, or that they do, but managed by an external device. They could also be part of a multisig arrangement, in which case it does have some keys and needs to generate (multisig) change addresses, but shouldn't generate "solo" keys.

If I understand correctly, the scope of this PR is wallets where keys are managed externally, e.g. a hardware wallet. Would such a wallet have it's own UI that uses the RPC, or would the user interact directly with the RPC somehow? In the first case, it would seem that the hardware wallet developer should just be careful about which RPC commands to call. Or am I missing something? Not that I'm against making it safer.

jonasschnelli commented:
Rebased and fixed.
Renamed `-disablehot` to `-disableprivatekeys`

Since we have no dynamic wallet creation (something like #10740), disabling private keys is only possible during wallet creation and with the `-disableprivatekeys` option.

Use case:
* Splitting the private keys from the wallet (hardware wallets, cold storage, use wallet as pure index), one wants to make sure, fundrawtransaction, etc. won't accidentally use an internal private key as change address or something similar

jonasschnelli commented:
Rebased (and fixed travis issue)

jonasschnelli commented:
Rebased.

jonasschnelli commented:
Had to rebase after #10244 (non trivial).
Glad if utACKers can review again.

kallewoof commented:
Late so can be safely ignored, but we say `importprivkey` so maybe `disableprivkeys` for consistency? Will review.

kallewoof commented on src/wallet/wallet.h:
```diff
@@ -1149,6 +1150,15 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
+    /** check if a certain wallet flag is set */
+    bool IsWalletFlagSet(uint64_t flag);
+
+    /** overwrite all flags by the given unit64_t */
```
Typo: `uint64_t`

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -1483,6 +1483,25 @@ bool CWallet::IsHDEnabled() const
+        throw std::runtime_error(std::string(__func__) + ": writing wallet flags failed");
+}
+
+bool CWallet::IsWalletFlagSet(uint64_t flag)
```
μNit: Why not just `GetWalletFlag`?

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -1495,11 +1497,17 @@ bool CWallet::IsWalletFlagSet(uint64_t flag)
     m_wallet_flags = overwriteFlags;
+    if ( (overwriteFlags >> 32 | g_known_wallet_flags >> 32) != (g_known_wallet_flags >> 32) ) {
+        // contains unkown non-tolerable wallet flags
+        return false;
```
I wonder if this should `throw` here instead of being a `can-fail` method.

jonasschnelli replied:
I think a specific error handling makes more sense (see error handling in walletdb)

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        if (gArgs.GetBoolArg("-disableprivatekeys", DEFAULT_DISABLE_PRIVKEY_WALLET)) {
+            walletInstance->AddWalletFlag(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
+        }
+        else {
```
Nit: `} else {`

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
 
         walletInstance->SetBestChain(chainActive.GetLocator());
+    }
+    else if (gArgs.IsArgSet("-disableprivatekeys")) {
```
Nit: `} else if (...) {`

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        LOCK(walletInstance->cs_wallet);
+        if ((!walletInstance->mapKeys.empty() || !walletInstance->mapCryptedKeys.empty()) && disablePrivKeys) {
+            InitError(strprintf(_("Error loading %s: You can't disable private keys if your wallet already contains private keys"), walletFile));
+            return NULL;
```
I think we prefer `nullptr` over `NULL` these days.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        }
+        if (walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !disablePrivKeys) {
+            InitError(strprintf(_("Error loading %s: You can't enable private keys once you have initialized a wallet with disabled private keys (use -disableprivatekeys instead)"), walletFile));
+            return NULL;
```
I think we prefer `nullptr` over `NULL` these days.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        /* make sure we don't have private keys */
+        if (walletInstance->GetKeyPoolSize()) {
+            InitError(strprintf(_("Error loading %s: Private keys present in a wallet where private keys are disabled"), walletFile));
+            return NULL;
```
I think we prefer `nullptr` over `NULL` these days.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4014,18 +4036,44 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+            return NULL;
+        }
+    }
+    else if (walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
Nit: `} else if ...`

kallewoof commented on src/wallet/wallet.h:
```diff
@@ -1157,8 +1167,9 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
 
-    /** overwrite all flags by the given unit64_t */
-    void SetWalletFlags(uint64_t overwriteFlags, bool memOnly);
+    /** overwrite all flags by the given unit64_t
```
Typo: `uint64_t`

kallewoof commented on src/wallet/walletdb.cpp:
```diff
@@ -568,10 +571,14 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
-                else
-                {
+                }
+                else if(strType == "flags") {
```
Nit: `} else`

kallewoof commented on src/wallet/walletdb.cpp:
```diff
@@ -568,10 +571,14 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+                    // reading the wallet flags can only fail if unknown flags are present
+                    result = DBErrors::TOO_NEW;
+                }
+                else {
```
Nit: `} else`

kallewoof commented on test/functional/wallet_disableprivkeys.py:
```diff
@@ -0,0 +1,58 @@
+
+        self.nodes[1].importpubkey(n0pubkR) #TODO switch to importmulti
+        self.nodes[1].importpubkey(n0pubkC)
+        assert_equal(self.nodes[1].getbalance("*", 1, True), 10)
```
Unsure why you're explicitly giving the `"*", 1, True` and not just doing `getbalance()` here.

kallewoof commented on test/functional/wallet_disableprivkeys.py:
```diff
@@ -0,0 +1,58 @@
+        assert_equal(self.nodes[1].getbalance("*", 1, True), 10)
+        rawtx = self.nodes[1].createrawtransaction([], {n0addr: 1.0})
+        frawtx= self.nodes[1].fundrawtransaction(rawtx, {"changeAddress": n0addrC, "includeWatching": True})
+        assert_equal(self.nodes[1].signrawtransactionwithwallet(frawtx['hex'])['complete'], False)
```
Maybe add a test that ensures it can `signrawtransactionwithkeys` (disabling privkeys should still let you sign, if you provide the keys, I think?).

kallewoof reviewed (commented):
utACK 74d25a289ea040ed33e0aaed120e77f3be7e5311

Note 3rd commit message needs to be updated (still says `disablehot`).

Edit: I know I'm late to the party, so take my nits with a grain of salt.

Sjors commented:
Concept ACK on the use of flags that are stored in the wallet (I think I overlooked that the first time I reviewed). This prevents accidentally disabling the feature. I also think it's a safe choice that you can't enable this feature on a wallet that already has private keys.

There are some problems in a multi wallet environment though.

It correctly remembers which wallets have disabled private keys, disabling the payment request button in QT where appropriate. So that's good.

But if you try to load an existing normal wallet and create a new `-disableprivatekeys` wallet at the same time, you'll get an error upon launch "You can't disable private keys if your wallet already contains private keys".

One solution would be to explain in documentation that in order to create such a wallet you can only have one `-wallet` argument. But I think a better approach is to create a new `-createwallet` `RPC` which takes `disableprivatekeys` as an argument.

jonasschnelli committed:
Add disable privatekeys option to createwallet

jonasschnelli committed:
[QA] add createwallet disableprivatekey test

jonasschnelli commented:
Overhauled this PR.
* Removed the global `-disableprivatkeys` argument.
* Added an `options` object to `createwallet` with currently a single boolean `disableprivatekeys`.

Sjors commented:
Much better!

Some stray files made it into the PR though (Bitcoin-Core.*).

jonasschnelli commented:
Thanks @Sjors. Removed the accidentally added IDE files.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3102,12 +3114,16 @@ UniValue loadwallet(const JSONRPCRequest& request)
             "1. \"wallet_name\"    (string, required) The name for the new wallet. If this is a path, the wallet will be created at the path location.\n"
+            "2. options          (json, optional)\n"
+            "  {\n"
+            "     \"disableprivatekeys\": <state>,         (boolean, optional, default: false) Disable the possibility of private keys (only watchonlys are possible in this mode).\n"
```
I was under the impression that named arguments is preferable to `options` object.

jnewbery replied:
I agree. Named arguments are much preferable. Using named arguments provides type- and error- checking for free, and simplify/flatten the interface

For example, if I typo the an option `disableprivtaekeys` here, this implementation will silently ignore the option. That's not possible with named arguments.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1513,6 +1515,31 @@ bool CWallet::IsHDEnabled() const
+bool CWallet::SetWalletFlags(uint64_t overwriteFlags, bool memonly)
+{
+    m_wallet_flags = overwriteFlags;
+    if ( (overwriteFlags >> 32 | g_known_wallet_flags >> 32) != (g_known_wallet_flags >> 32) ) {
```
`if ((overwriteFlags & ^g_known_wallet_flags) >> 32) {` seems slightly more readable to me.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4164,18 +4205,34 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
+    } else if (walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+        LOCK(walletInstance->cs_wallet);
+        if (!walletInstance->mapKeys.empty() || !walletInstance->mapCryptedKeys.empty()) {
+            InitError(strprintf(_("Error loading %s: Private keys detected in wallet with disabled private keys"), walletFile));
```
This seems a little dangerous. If we get into this condition for any reason, it'll be impossible for the user to load the wallet and retrieve their private keys.

Sjors replied:
Sounds like that should be a warning, not an error.

jnewbery commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -99,6 +99,9 @@ void ReceiveCoinsDialog::setModel(WalletModel *_model)
             ui->useBech32->setCheckState(Qt::Unchecked);
         }
+
+        // eventually disable the main receive button if private key operations are disabled
```
How does this interact with multiwallet? Loading/unloading wallets at run time?

Sjors replied:
I still think multiwallet should use one window per wallet rather than a dropdown inside a single window. Pending that, receive screen should at least be visible if any of the wallets aren't read-only. The receive button itself would have to be updated when the user switches between wallets.

Another question is: provided there's enough warnings (including about scams), why shouldn't a user be able to receive money on a watch-only address?

jonasschnelli replied:
@Sjors comment is orthogonal.

@jnewbery: this seems to work perfectly fine with mtulitwallet. Not sure about unloading... but I guess unloading does also call `setModel()`... or can you be more specific why this should be a problem?

jnewbery replied:
Yes, you're right that this works fine with multiwallet.

It would be nice if there was help text saying that the 'receive' button is disabled because the wallet has private keys disabled (and so can't generate new receiving addresses). However, that could be added in a separate PR.

jnewbery commented on test/functional/wallet_multiwallet.py:
```diff
@@ -151,7 +151,9 @@ def run_test(self):
         assert_equal(w4.getbalance(), 0)
 
-        w1.sendtoaddress(w2.getnewaddress(), 1)
+        w2addrR = w2.getnewaddress()
```
nit: I'm not sure what the `R` is for here. In general, prefer to use `lower_case_with_underscores` for [variable names in Python](https://www.python.org/dev/peps/pep-0008/#id45).

jnewbery commented on test/functional/wallet_multiwallet.py:
```diff
@@ -231,8 +233,14 @@ def run_test(self):
-
         assert new_wallet_name in self.nodes[0].listwallets()
 
+        self.log.info("Test disableprivatekeys creation.")
```
It's not obvious to me why this test is in `wallet_multiwallet.py`. What does it have to do with multiwallet?

jonasschnelli replied:
I guess I placed it there because this was the only test that used `createwallet`. But right, I shall move it to a new test.

jnewbery reviewed (commented):
Concept ACK. I can run the new functional test, but I wasn't able to get the receive tab to disappear in qt.

A couple of high-level things that I think could improve this PR:
- some unit test of the new wallet/walletdb functionality.
- update `getwalletinfo` to display whether the wallet has private keys disabled.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -4183,18 +4205,34 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
+        if ((wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+            //selective allow to set flags
+            walletInstance->AddWalletFlag(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
+        }
```
Commit "Add per option to disable private keys during internal wallet creation"

nit `} else {`.

jonasschnelli committed:
QA: Fix bug in -usecli logic that converts booleans to non-lowercase strings

jonasschnelli commented:
Fixed @jnewbery s points.
* Added `privatekeys_enabled` to `getwalletinfo`
* Added unit test and own rpc test

jnewbery reviewed (commented):
I tested 4c2b3fc3b and it looks great. A few nits inline.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1513,6 +1513,25 @@ bool CWallet::IsHDEnabled() const
 
+void CWallet::AddWalletFlag(uint64_t flags)
+{
+    m_wallet_flags |= flags;
```
Fixed.

jonasschnelli commented on src/wallet/wallet.h:
```diff
@@ -1198,6 +1209,16 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     /** Whether a given output is spendable by this wallet */
     bool OutputEligibleForSpending(const COutput& output, const CoinEligibilityFilter& eligibility_filter) const;
+
+    /** set a single wallet flag */
```
Agree... changed from AddWalletFlag to SetWalletFlag

jonasschnelli commented:
Fixed remaining points reported by @jnewbery and @kallewoof

jnewbery commented on src/wallet/walletdb.cpp:
```diff
@@ -570,10 +578,13 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+                } else if(strType == "flags") {
+                    // reading the wallet flags can only fail if unknown flags are present
+                    result = DBErrors::TOO_NEW;
+                }
```
nit: `} else {`

jnewbery commented on src/wallet/walletdb.cpp:
```diff
@@ -510,7 +510,15 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+                strErr = "Error reading wallet database: Unknown non-tolerable wallet flags found";
+                return false;
+            }
+        }
```
nit: `} else if (...) {`

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3132,7 +3152,7 @@ UniValue createwallet(const JSONRPCRequest& request)
     }
 
-    std::shared_ptr<CWallet> const wallet = CWallet::CreateWalletFromFile(wallet_name, fs::absolute(wallet_name, GetWalletDir()));
+    std::shared_ptr<CWallet> const wallet = CWallet::CreateWalletFromFile(wallet_name, fs::absolute(wallet_name, GetWalletDir()), (disable_privatekeys ? WALLET_FLAG_DISABLE_PRIVATE_KEYS : 0));
```
This introduces a new build warning for me:

```
wallet/rpcwallet.cpp: In function ‘UniValue createwallet(const JSONRPCRequest&)’:
wallet/rpcwallet.cpp:3155:152: warning: enumeral and non-enumeral type in conditional expression [-Wextra]
     std::shared_ptr<CWallet> const wallet = CWallet::CreateWalletFromFile(wallet_name, fs::absolute(wallet_name, GetWalletDir()), (disable_privatekeys ? WALLET_FLAG_DISABLE_PRIVATE_KEYS : 0));
                                                                                                                                                        ^
```

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -113,6 +113,16 @@ constexpr OutputType DEFAULT_ADDRESS_TYPE{OutputType::P2SH_SEGWIT};
+    WALLET_FLAG_DISABLE_PRIVATE_KEYS = (1ULL << 32),
+};
+
+static const uint64_t g_known_wallet_flags = WALLET_FLAG_DISABLE_PRIVATE_KEYS;
```
nit: use constexpr

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1513,6 +1515,33 @@ bool CWallet::IsHDEnabled() const
+        return false;
+    }
+    if (!memonly && !WalletBatch(*database).WriteWalletFlags(m_wallet_flags))
+        throw std::runtime_error(std::string(__func__) + ": writing wallet flags failed");
```
nit: if block should have braces.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4164,18 +4207,32 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
+            // generate a new seed
+            CPubKey seed = walletInstance->GenerateNewSeed();
+            if (!walletInstance->SetHDSeed(seed))
+                throw std::runtime_error(std::string(__func__) + ": Storing HD seed failed");
```
nit: if block should use braces

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4164,18 +4207,32 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
         }
 
         walletInstance->ChainStateFlushed(chainActive.GetLocator());
+    } else if ((wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
nit: no need for double parens

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4164,18 +4207,32 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
+    } else if (walletInstance->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+        LOCK(walletInstance->cs_wallet);
+        if (!walletInstance->mapKeys.empty() || !walletInstance->mapCryptedKeys.empty()) {
+            InitWarning(strprintf(_("Error loading %s: Private keys detected in wallet with disabled private keys"), walletFile));
```
nit: change log to:

`"Warning: Private keys detected in wallet {%s} with disabled private keys"), walletFile));`

(since this is now a warning instead of an error)

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2970,19 +2982,20 @@ static UniValue getwalletinfo(const JSONRPCRequest& request)
+            "  \"paytxfee\": x.xxxx,                (numeric) the transaction fee configuration, set in " + CURRENCY_UNIT + "/kB\n"
+            "  \"hdseedid\": \"<hash160>\"            (string, optional) the Hash160 of the HD seed (only present when HD is enabled)\n"
+            "  \"hdmasterkeyid\": \"<hash160>\"       (string, optional) alias for hdseedid retained for backwards-compatibility. Will be removed in V0.18.\n"
+            "  \"privatekeys_enabled\": true|false  (boolean) false if privatekeys are disabled for this wallet (enforced watch-only wallet)\n"
```
nit: `private_keys_enabled`

jnewbery reviewed (commented):
A few more nits. Sorry!

jonasschnelli commented:
Fixed @jnewbery's its.

laanwj commented:
utACK 5caf38f42f34b2aec7440dec752c10856e5d85a7

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -1513,6 +1515,34 @@ bool CWallet::IsHDEnabled() const
+{
+    LOCK(cs_wallet);
+    m_wallet_flags = overwriteFlags;
+    if ( (overwriteFlags >> 32 | g_known_wallet_flags >> 32) != (g_known_wallet_flags >> 32) ) {
```
Code style is a bit funky here.

kallewoof replied:
Personally think this definitely needs parens to clarify priority, but maybe I'm just easily confused.

jonasschnelli replied:
Fixed

laanwj commented:
@jonasschnelli didn't you accidentally remove the logic so that only
changes to the upper 32 bits are checked?

On Wed, Jul 4, 2018 at 5:01 PM, Jonas Schnelli <notifications@github.com>
wrote:

> *@jonasschnelli* commented on this pull request.
> ------------------------------
>
> In src/wallet/wallet.cpp
> <https://github.com/bitcoin/bitcoin/pull/9662#discussion_r200152951>:
>
> > +    LOCK(cs_wallet);
> +    m_wallet_flags |= flags;
> +    if (!WalletBatch(*database).WriteWalletFlags(m_wallet_flags))
> +        throw std::runtime_error(std::string(__func__) + ": writing wallet flags failed");
> +}
> +
> +bool CWallet::IsWalletFlagSet(uint64_t flag)
> +{
> +    return (m_wallet_flags & flag);
> +}
> +
> +bool CWallet::SetWalletFlags(uint64_t overwriteFlags, bool memonly)
> +{
> +    LOCK(cs_wallet);
> +    m_wallet_flags = overwriteFlags;
> +    if ( (overwriteFlags >> 32 | g_known_wallet_flags >> 32) != (g_known_wallet_flags >> 32) ) {
>
> Fixed
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9662#discussion_r200152951>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHutoKKghkVC66tkbFR3O1hWX2N0O9Vks5uDNjCgaJpZM4Lz2fK>
> .
>

Sjors commented:
Lightly tested by creating a wallet via RPC, seeing it appear in the GUI, importing a watch-only address and sending some coins to it from a normal wallet. That money doesn't show up, even after confirmation, until I restart. Perhaps it's an unrelated multiwallet issue though.

@jnewbery said somewhere above:

> Named arguments are much preferable. Using named arguments provides type- and error- checking for free, and simplify/flatten the interface
> 
> For example, if I typo the an option disableprivtaekeys here, this implementation will silently ignore the option. That's not possible with named arguments.

`disable_private_keys` is still positional, which would be painful to change later...

This GUI behavior is confusing - because it suggests the wallet isn't HD, even though it may or may not be -  but I'm fine with addressing it later:

<img width="100" alt="gui" src="https://user-images.githubusercontent.com/10217/42517736-b6290380-8460-11e8-8fa0-cea1b76c2e7e.png">

jonasschnelli commented:
Rebased and fixed the upper-bits restriction (for the wallet flags) found by @laanwj (https://github.com/bitcoin/bitcoin/pull/9662#issuecomment-402735925).

Sjors commented:
Unhappy Travis for the Darwin machine:
<img width="1839" alt="schermafbeelding 2018-07-12 om 13 13 49" src="https://user-images.githubusercontent.com/10217/42629954-73b28e10-85d5-11e8-98e6-eed75f21bc06.png">

I these warnings too when locally on macOS:
<img width="1096" alt="schermafbeelding 2018-07-12 om 13 20 39" src="https://user-images.githubusercontent.com/10217/42630263-703a1eb4-85d6-11e8-8ecd-c7af83143f0f.png">

jonasschnelli commented:
Fixed the rebase issue (the warning; requires `cs_KeyStore` instead of `cs_wallet` now).
